v2/relay/constraints.go

```diff
@@ -0,0 +1,153 @@
+package relay
+
+import (
+	"container/list"
+	crand "crypto/rand"
+	"encoding/binary"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	asnutil "github.com/libp2p/go-libp2p-asn-util"
+	"github.com/libp2p/go-libp2p-core/peer"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+var validity = 30 * time.Minute
+
+var (
+	errTooManyReservations        = errors.New("too many reservations")
+	errTooManyReservationsForPeer = errors.New("too many reservations for peer")
+	errTooManyReservationsForIP   = errors.New("too many peers for IP address")
+	errTooManyReservationsForASN  = errors.New("too many peers for ASN")
+)
+
+type listEntry struct {
+	t time.Time
+}
+
+// constraints implements various reservation constraints
+type constraints struct {
+	rc *Resources
+
+	mutex sync.Mutex
+	rand  rand.Rand
+	total *list.List
```

Comment:
Any reason not to just use slices?

Reply:
Because I said lists... sorry. I've been programming in lisp too much lately. A slice is probably much simpler and more performant.

Reply:
Yeah, and I've clearly not been thinking clearly. Of course a slice is much better.